### PR TITLE
Fix default values

### PR DESCRIPTION
**Pull request by ltorroba**
This PR fixes problems when making requests to the server before the attributes have been populated. It also includes some tests for the new functionality.

Some minor bug fixes were also made (especially with our use of shelve).

### CODE REVIEW

**Comment by ltorroba** [comment]
@james-od Ready for review

File: flaskapp/flaskapp.py [hunk 1]
```diff
@@ -20,7 +20,8 @@ def get_cache():
     It can store arbitrary objects.
     """
     if not hasattr(g, 'cache'):
-        g.cache = shelve.open(app.config['SHELVE_FILENAME'])
+        g.cache = shelve.open(app.config['SHELVE_FILENAME'], protocol=2,
```
Comment (james-od) [inline flaskapp/flaskapp.py:23]: Don't understand this bit but assume you know what you're doing 😛

**Review by james-od — APPROVED** [review]
Just don't understand that one line, rest looks great!

File: flaskapp/flaskapp.py (hunk 1 above)
Comment (ltorroba) [reply flaskapp/flaskapp.py:23]: Yeah, a bit of context for that line: strictly the `protocol` wasn't necessary as shelve uses protocol 2 by default when running Python 2.7, but if you are running Python 3 it seems to cause some problems.

Technically we don't need to support Python 3, but as it's the only line (as far as I know) that needs it, I figured I might as well add it.

Can definitely get it removed though if you think it's better. 👍

Comment (james-od) [reply flaskapp/flaskapp.py:23]: Nah dude just wondered what it did - definitely good to keep it

**Pull request merged**